numba/byteflow.py

```diff
@@ -0,0 +1,1139 @@
+import logging
+from collections import namedtuple, defaultdict
+
+from numba.utils import UniqueDict, PYVERSION
+from numba.controlflow import NEW_BLOCKERS, CFGraph
+from numba.ir import Loc
+from numba.errors import UnsupportedError
+
+
+_logger = logging.getLogger(__name__)
+# logging.basicConfig(level=logging.DEBUG)
+
+
+class _lazy_pformat(object):
+    def __init__(self, *args, **kwargs):
+        self.args = args
+        self.kwargs = kwargs
+
+    def __str__(self):
+        return pformat(*self.args, **self.kwargs)
+
+
+class Flow(object):
+    """Data+Control Flow analysis.
+
+    Simulate execution to recover dataflow and controlflow information.
+    """
+    def __init__(self, bytecode):
+        _logger.debug("bytecode dump:\n%s", bytecode.dump())
+        self._bytecode = bytecode
+        self.block_infos = UniqueDict()
+
+    def run(self):
+        firststate = State(bytecode=self._bytecode, pc=0, nstack=0,
+                           blockstack=())
+        runner = Runner(debug_filename=self._bytecode.func_id.filename)
+        runner.pending.append(firststate)
+
+        # Enforce unique-ness on initial PC to avoid re-entering the PC with
+        # a different stack-depth. We don't know if such a case is ever
+        # possible, but no such cases has been encountered in our tests.
+        first_encounter = UniqueDict()
+        # Loop over each pending state at a initial PC.
+        # Each state is tracing a basic block
+        while runner.pending:
+            _logger.debug("pending: %s", runner.pending)
+            state = runner.pending.pop()
+            if state not in runner.finished:
+                _logger.debug("stack: %s", state._stack)
+                first_encounter[state.pc_initial] = state
+                # Loop over the state until it is terminated.
+                while True:
+                    runner.dispatch(state)
+                    # Terminated?
+                    if state.has_terminated():
+                        break
+                    # Not yet?
+                    else:
+                        state.advance_pc()
+                        # Must the new PC be a new block?
+                        if self._is_implicit_new_block(state):
+                            state.split_new_block()
+                            break
+                _logger.debug("end state. edges=%s", state.outgoing_edges)
+                runner.finished.add(state)
+                out_states = state.get_outgoing_states()
+                runner.pending.extend(out_states)
+
+        # Complete controlflow
+        self._build_cfg(runner.finished)
+        # Prune redundant PHI-nodes
+        self._prune_phis(runner)
+        # Post process
+        for state in sorted(runner.finished, key=lambda x: x.pc_initial):
+            self.block_infos[state.pc_initial] = si = adapt_state_infos(state)
+            _logger.debug("block_infos %s:\n%s", state, si)
+
+        assert self.block_infos
```

Comment:
Given this will propagate to user land, perhaps check and raise a Numba exception instead?

Reply:
this can actually be removed. added to debug.